src/services/qix/session.js

```diff
@@ -133,6 +170,143 @@ class Session {
+  * @returns {Object} Returns a promise instance.
+  */
+  restoreDoc(sessionState, restoredApis) {
+    const app = this.apis.getApisByType('Doc').pop();
+    if (!app) {
+      return this.Promise.resolve();
+    }
+
+    const attached = sessionState === 'SESSION_ATTACHED';
+    const method = attached ? 'GetActiveDoc' : 'OpenDoc';
+    const params = attached ? [] : [this.appId, '', '', '', !!this.noData];
+    const request = {
+      method,
+      handle: -1,
+      params,
+    };
+
+    return this.rpc.send(request)
+      .then((response) => {
+        if (response.error) {
+          return this.Promise.reject(new Error(response.error.message));
+        }
+
+        const handle = response.result.qReturn.qHandle;
+        app.api.handle = handle;
+        restoredApis.add(handle, app.api);
+        return this.Promise.resolve(app.api);
+      });
+  }
+
+  /**
+  * Function used to restore the APIs on the doc.
+  * @param {Object} doc - The doc API on which the APIs we want to restore exist.
+  * @param {Object} restoredApis - The API-cache containing restored APIs.
+  * @param {Array} closed - A list of APIs that has been closed when resuming the session.
+  * @returns {Object} Returns a promise instance.
+  */
+  restoreDocObjects(doc, restoredApis, closed) {
+    const tasks = [];
+
+    const entries = this.apis.getApis().filter(entry => entry.api.type !== 'Global' && entry.api.type !== 'Doc');
+
+    entries.forEach((entry) => {
+      const api = entry.api;
+      const method = api.getMethodName;
+
+      if (!method) {
+        closed.push(api);
+      } else {
+        const request = this.rpc.send({
+          method,
+          handle: doc.handle,
+          params: [api.id],
+        })
+        .then((response) => {
+          if (response.error || !response.result.qReturn.qHandle) {
+            closed.push(api);
+          } else {
+            api.handle = response.result.qReturn.qHandle;
+            restoredApis.add(api.handle, api);
+          }
+        });
+        tasks.push(request);
+      }
+    });
+    return Promise.all(tasks);
+  }
+
+  /**
+  * Function used to resume the session
+  * @param {Boolean} onlyIfAttached - if true, resume only if the session was re-attached.
+  * @returns {Object} Returns a promise instance.
+  */
+  resume(onlyIfAttached) {
+    if (!this.suspended) {
+      return this.Promise.reject(new Error('Not suspended'));
+    }
+
+    this.unregisterRpcListeners();
+    const restoredApis = new ApiCache();
+    const closed = [];
+
+    return this.rpc.reopen(ON_ATTACHED_TIMEOUT)
+      .then((sessionState) => {
+        if (sessionState === 'SESSION_CREATED' && onlyIfAttached) {
+          return this.Promise.reject(new Error('Not attached'));
+        }
+        return this.Promise.resolve(sessionState);
+      })
+      .then(sessionState => this.restoreGlobal(restoredApis)
+        .then(() => this.restoreDoc(sessionState, restoredApis))
```

Comment:
Since we're always doing `GetActiveDoc` first (if we use the approach I suggested in `restoreDoc()`), there's no need to do this inside the other `.then(...)` call (we no longer need `sessionState`), move it out to the outer promise chain.